src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultHtmlGenerator.cs

```diff
@@ -746,18 +722,137 @@ public virtual TagBuilder GenerateAntiForgery([NotNull] ViewContext viewContext)
+                    if (rawValue is IEnumerable<SelectListItem>)
+                    {
+                        // This ViewData item contains the fallback selectList collection for GenerateSelect().
+                        // Do not try to use this collection.
+                        rawValue = null;
+                    }
+                }
+                else
+                {
+                    // <select/>, Html.DropDownListFor() and Html.ListBoxFor() helper case. Do not use ViewData.
+                    rawValue = modelExplorer.Model;
+                }
+
+                if (rawValue == null)
+                {
+                    return null;
+                }
+            }
+
+            // Convert raw value to a collection.
+            IEnumerable rawValues;
+            if (allowMultiple)
+            {
+                rawValues = rawValue as IEnumerable;
+                if (rawValues == null || rawValues is string)
+                {
+                    throw new InvalidOperationException(
+                        Resources.FormatHtmlHelper_SelectExpressionNotEnumerable(nameof(expression)));
+                }
+            }
+            else
+            {
+                rawValues = new[] { rawValue };
+            }
+
+            modelExplorer = modelExplorer ??
+                ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+
+            var enumNames = modelExplorer.Metadata.EnumNamesAndValues;
+            var isTargetEnum = modelExplorer.Metadata.IsEnum;
+            var innerType =
+                Nullable.GetUnderlyingType(modelExplorer.Metadata.ModelType) ?? modelExplorer.Metadata.ModelType;
+
+            // Convert raw value collection to strings.
+            var currentValues = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var value in rawValues)
+            {
+                // Add original or converted string.
+                var stringValue = (value as string) ?? Convert.ToString(value, CultureInfo.CurrentCulture);
+
+                // Do not add simple names of enum properties here because whitespace isn't relevant for their binding.
+                // Will add matching names just below.
+                if (enumNames == null || !enumNames.ContainsKey(stringValue.Trim()))
+                {
+                    currentValues.Add(stringValue);
+                }
+
+                // Remainder handles isEnum cases. Convert.ToString() returns field names for enum values but select
+                // list may (well, should) contain integer values.
+                var enumValue = value as Enum;
+                if (isTargetEnum && enumValue == null && value != null)
+                {
+                    var valueType = value.GetType();
+                    if (typeof(long).IsAssignableFrom(valueType) || typeof(ulong).IsAssignableFrom(valueType))
```

Comment:
that's the reason for `IsAssignableFrom()`.  for example the tests use `int` and conversion to the appropriate `TEnum` works fine.